File: bindings/python/ranked_belief/examples/hidden_markov.py
"""Hidden Markov model example expressed with the Python bindings."""

from __future__ import annotations

from collections.abc import Iterable, Sequence
from functools import lru_cache

from .. import Rank, RankingFunctionAny
from ..dsl import either_or, merge_apply, normal_exceptional, observe, take_n

__all__ = ["hmm", "sequence_likelihoods"]

State = str
Observation = str
Path = tuple[tuple[State, Observation | None], ...]


@lru_cache(maxsize=None)
def init() -> RankingFunctionAny:
    """Initial distribution over weather states."""

    return either_or("rainy", "sunny")


def trans(state: State) -> RankingFunctionAny:
    """Transition model (probabilistic ranking)."""

    if state == "rainy":
        return normal_exceptional("rainy", "sunny", Rank.from_value(2))
    return normal_exceptional("sunny", "rainy", Rank.from_value(2))


def emit(state: State) -> RankingFunctionAny:
    """Emission model linking latent state to umbrella observation."""

    if state == "rainy":
        return normal_exceptional("yes", "no", Rank.from_value(1))
    return normal_exceptional("no", "yes", Rank.from_value(1))


def _prepend_step(path: Path, state: State, observation: Observation) -> RankingFunctionAny:
    new_path = path + ((state, observation),)
    return RankingFunctionAny.singleton(new_path)


def _advance(path: Path, observation: Observation) -> RankingFunctionAny:
    last_state = path[-1][0]

    def bind_state(next_state: State) -> RankingFunctionAny:
        consistent_emissions = observe(
            emit(next_state),
            lambda value: value == observation,
        )
        return merge_apply(
            consistent_emissions,
            lambda _matched: _prepend_step(path, next_state, observation),
        )

    return merge_apply(trans(last_state), bind_state)


def hmm(observations: Sequence[Observation]) -> RankingFunctionAny:
    """Compute the ranking over state/observation paths matching ``observations``."""

    def initialise(state: State) -> RankingFunctionAny:
        start: Path = ((state, None),)
        return RankingFunctionAny.singleton(start)

    ranking = merge_apply(init(), initialise)
    for obs in observations:
        ranking = merge_apply(ranking, lambda path, obs=obs: _advance(path, obs))
    return ranking.map(lambda path: tuple(path[1:]))


def sequence_likelihoods(observations: Iterable[Observation], *, limit: int = 5) -> list[tuple[Path, Rank]]:
    """Return the most plausible latent state sequences for ``observations``."""

    return take_n(hmm(tuple(observations)), limit)

File: bindings/rankedBeliefR/inst/examples/hidden_markov.R
# Hidden Markov model example (full port from Python)
#
# Weather HMM: rainy/sunny states, umbrella yes/no observations

library(rankedBeliefR)

# Encode states and emissions as integers
RAINY <- 1L
SUNNY <- 2L
YES <- 1L
NO <- 2L

state_name <- c("rainy", "sunny")
emission_name <- c("yes", "no")

# Initial distribution: equal plausibility
init_distribution <- function() {
  rb_from_array_int(values = c(RAINY, SUNNY), ranks = c(0, 0))
}

# Transition model
trans <- function(state) {
  if (state == RAINY) {
    # rainy -> rainy (normal), rainy -> sunny (exceptional, rank 2)
    rb_from_array_int(values = c(RAINY, SUNNY), ranks = c(0, 2))
  } else {
    # sunny -> sunny (normal), sunny -> rainy (exceptional, rank 2)
    rb_from_array_int(values = c(SUNNY, RAINY), ranks = c(0, 2))
  }
}

# Emission model
emit <- function(state) {
  if (state == RAINY) {
    # yes (normal), no (exceptional, rank 1)
    rb_from_array_int(values = c(YES, NO), ranks = c(0, 1))
  } else {
    # no (normal), yes (exceptional, rank 1)
    rb_from_array_int(values = c(NO, YES), ranks = c(0, 1))
  }
}

# Simple HMM forward pass (behaviorally equivalent but not fully lazy)
# Returns ranking over final states given observations
hmm <- function(observations) {
  # Start with initial distribution
  current_states <- rb_take_n_int(init_distribution(), 10)
  
  # For each observation, update beliefs
  for (obs in observations) {
    next_scenarios <- list()
    
    for (i in seq_len(nrow(current_states))) {
      state <- current_states$value[i]
      state_rank <- current_states$rank[i]
      
      # Transition to next states
      trans_df <- rb_take_n_int(trans(state), 10)
      
      for (j in seq_len(nrow(trans_df))) {
        next_state <- trans_df$value[j]
        trans_rank <- trans_df$rank[j]
        
        # Check emission consistency
        emit_df <- rb_take_n_int(emit(next_state), 10)
        matching <- emit_df[emit_df$value == obs, ]
        
        if (nrow(matching) > 0) {
          emit_rank <- matching$rank[1]
          total_rank <- state_rank + trans_rank + emit_rank
          next_scenarios[[length(next_scenarios) + 1]] <- 
            list(state = next_state, rank = total_rank)
        }
      }
    }
    
    # Aggregate and build next distribution
    if (length(next_scenarios) > 0) {
      states_vec <- sapply(next_scenarios, function(s) s$state)
      ranks_vec <- sapply(next_scenarios, function(s) s$rank)
      temp <- rb_from_array_int(values = as.integer(states_vec), ranks = ranks_vec)
      current_states <- rb_take_n_int(temp, 100)
      rb_free(temp)
    }
  }
  
  current_states
}

run_example <- function() {
  cat("Hidden Markov model example\n")
  cat("Observations: umbrella=yes, umbrella=yes\n\n")
  
  result <- hmm(c(YES, YES))
  result <- result[order(result$rank), ]
  top <- head(result, 5)
  
  cat("Top state sequences:\n")
  for (i in seq_len(nrow(top))) {
    state_label <- state_name[top$value[i]]
    cat(sprintf("%d. state=%s rank=%g\n", i, state_label, top$rank[i]))
  }
}

if (!exists("SKIP_EXAMPLE_RUN")) run_example()
